packages/core-state/src/wallets/wallet-repository.ts

```diff
@@ -139,15 +139,25 @@ export class WalletRepository implements Contracts.State.WalletRepository {
     }
 
     public forgetByIndex(indexName: string, key: string): void {
+        const forgottenWallet = this.getIndex(indexName).get(key);
+        for (const index of Object.values(this.indexes)) {
+            index.entries().forEach(([name, wallet]) => {
+                if (wallet.publicKey === forgottenWallet?.publicKey) {
+                    index.forget(name);
+                }
+            });
+        }
         this.getIndex(indexName).forget(key);
     }
 
+    // TODO: naming here implies the behaviour is different to reindexing
     public index(wallets: ReadonlyArray<Contracts.State.Wallet>): void {
         for (const wallet of wallets) {
             this.reindex(wallet);
         }
     }
 
+    // TODO: reindex gives the impression this wallet is already indexed - possibly change name?
```

Comment:
We should probably merge this into `index` and drop the method completely. We could need to change the signature of the `index` method to `public index(wallets: Contracts.State.Wallet | ReadonlyArray<Contracts.State.Wallet>): void` so we can pass in either a single wallet or array. Depending on how often `reindex` is called with a single wallet we might even drop it completely and simply pass in an array.

Reply:
Fixed here => b0d2d682c4d6356b1a0bc9250e9907396fa6ee10